File: apps/java/src/test/java/chapter08/Tables.java
package chapter08;

import chapter02.Row;
import chapter02.Table;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** テスト用の表を作り、列の値を読む。 */
final class Tables {
  private Tables() {}

  /** 1 つ目の引数を列名の並び、残りを行として、カンマ区切りの文字列から表を作る。空欄は欠損値。 */
  static Table table(String header, String... lines) {
    List<String> columns = List.of(header.split(","));
    List<Row> rows = Arrays.stream(lines).map(line -> row(columns, line)).toList();
    return new Table(columns, rows);
  }

  private static Row row(List<String> columns, String line) {
    String[] values = line.split(",", -1);
    Map<String, String> cells = new HashMap<>();
    for (int i = 0; i < columns.size(); i++) {
      cells.put(columns.get(i), values[i]);
    }
    return new Row(cells);
  }

  /** 数値の列の値を、上から順に並べる。欠損値があれば失敗する。 */
  static List<Double> numbers(Table table, String name) {
    return table.rows().stream().map(row -> row.number(name).orElseThrow()).toList();
  }

  /** 文字列の列の値を、上から順に並べる。 */
  static List<String> texts(Table table, String name) {
    return table.rows().stream().map(row -> row.text(name)).toList();
  }
}

File: apps/scala/src/test/scala/machinelearning/chapter08/Tables.scala
package machinelearning.chapter08

import machinelearning.chapter02.{Row, Table}

/** テスト用の表を作り、列の値を読む。 */
object Tables:

  /** 1 つ目の引数を列名の並び、残りを行として、カンマ区切りの文字列から表を作る。空欄は欠損値。 */
  def table(header: String, lines: String*): Table =
    val columns = header.split(",", -1).toVector
    Table(columns, lines.toVector.map(line => Row(columns.zip(line.split(",", -1).toVector).toMap)))

  /** 数値の列の値を、上から順に並べる。欠損値があれば失敗する。 */
  def numbers(table: Table, column: String): Vector[Double] =
    table.rows.map(_.number(column).getOrElse(throw IllegalStateException(s"欠損値です: $column")))

  /** 文字列の列の値を、上から順に並べる。 */
  def texts(table: Table, column: String): Vector[String] = table.rows.map(_.text(column))
